components/util/model.jsx

```diff
@@ -0,0 +1,60 @@
+import React from 'react';
+import { immutableMerge } from './index.js';
+
+export function Model({ initialState, methods }) {
+  const listeners = [];
+  const notifyListeners = nextState => listeners.forEach(fn => fn(nextState));
+
+  const model = {
+    state: Object.freeze(initialState),
+    subscribe: fn => listeners.push(fn),
+    unsubscribe: fn => listeners.splice(listeners.indexOf(fn), 1),
+  };
+
+  Object.keys(methods).forEach((method) => {
+    model[method] = (...args) => {
+      const updates = methods[method].apply(model, args);
+      if (updates && typeof updates === 'object') {
```

Comment:
Does it matter if `updates` is an array or an object?
both `typeof []` and `typeof {}` will return `'object'`